### ISSUE TITLE
working of feature/add-dropout

### ISSUE THREAD

**Issue by csetraynor** [dreamquark-ai/tabnet#290]
## Feature request
The dropout feature is already in development. I wanted to contact @Optimox to discuss if the feature works already or needs more development.

**What is the expected behavior?**

I expect that dropout is either an argument to TabNetRegressor/TabNetClassifier or clf.fit.
However, when I tried the development branch feature/add-dropout I found the error:
`__init__() got an unexpected keyword argument 'dropout'
`
I check the code but it would be useful to understand better why is failing.

**What is motivation or use case for adding/changing the behavior?**

Dropout can be useful in some situations.

**How should this be implemented in your opinion?**

I think that it would be awesome to have dropout in the new develop branch.

**Are you willing to work on this yourself?**
yes, I will keep digging the code or try to add the dropout layer in the `tabnet_network.py` file within the branch develop, but it would be useful to know @Optimox experience.

**Comment by eduardocarvp** [comment]
Hello @csetraynor ,

Indeed, the regressor/classifier classes do not have a dropout argument.
I do not remember if the paper mentions dropout, but where on the network do you think it would be interesting to add a dropout layer? It should not be too complicated on `tabnet_network.py`, like you said.

**Comment by csetraynor** [comment]
Hi @eduardocarvp  and thanks for the quick reply.

Yes, I considered it would make more sense to have Dropout after each Tabnet Step, for example, immediately after the ReLU activation. The other simpler option is to have Dropout after summing the Tabnet Steps just before the FC layer. To me, maybe it makes more sense to put it after each Tabnet Step immediately after the ReLU. Do you agree? 
Also, it would be nice to have an option to allow the use of Dropout in prediction time so we can approximately have a rough estimate of the dropout uncertainty for predictions.

**Comment by Optimox** [comment]
@csetraynor I think I tried adding dropout (which is not in the original paper) a long time ago.

I think it could be added very easily but I was worried at the time that dropout would be in conflict with the attention mechanism.
I'm not sure where it's best suited to use dropout (directly with input features? only for hidden predictions layers? also for attention layers?).

Have you tried some of those solutions on a specific dataset? Did you see any improvement?

**Comment by csetraynor** [comment]
@Optimox thanks for your reply. 
- I realised why the feature/add-dropout branch wasn't working for me! I forgot to modify the imports in tabnet_model to the name I gave to the folder of this branch but the feature is actually working so this thread may be closed if you feel fit.
- My goal is more obtaining some uncertainty in my predictions rather than improving performance, though improving performance is always a plus.
- Following your advice I just added a dropout directly with the input features that makes sense with the aims of the exercise.
- Now I found some improvement see below with dropout in the input features (after embbedings in the forward method of Tabnet).. But may be a fluke because this exercise is a simulation not a real-world benchmark. I will make real world benchmarks next week I can keep you updated.

![image](https://user-images.githubusercontent.com/33321804/117369652-82222c00-aec5-11eb-98c3-20b48e3b6288.png)

Thanks again!

**Comment by Optimox** [comment]
@csetraynor just be careful because the branch feature/add-dropout was a personal branch and I think there may be more differences than just adding dropout (I think I tried a bunch of different things). I need to be careful with my personal branches as they end up in public in the official repo :s

**Comment by csetraynor** [comment]
@Optimox thanks for the heads up. 
Also, sorry for the late reply. I didn't have a GPU at reach these days.
Yes, I am working with the new architecture, and I add a dropout layer to the input features. To do it, I am adding dropout after the embeddings layers. Currently, I have the dropout layer for the forward and the forward mask methods, for example:
```
    def forward(self, x):
        x = self.embedder(x)
        x = torch.nn.Dropout(self.dropout)(x)
        return self.tabnet(x)

    def forward_masks(self, x):
        x = self.embedder(x)
        x = torch.nn.Dropout(self.dropout)(x)
        return self.tabnet.forward_masks(x)
```

I used this approach for the census example dataset and I get more realistic results. See here:
![image](https://user-images.githubusercontent.com/33321804/118056585-826d6c00-b38a-11eb-97b0-bc4fb650daac.png)

Do you think any better place to add dropout? Do you believe that for dropout to the input features approach, I need the dropout for the forward_mask also or only for the forward methods?

**Comment by Optimox** [comment]
Hmm why do you say it's a more realistic result? It looks like dropout is very harmful in this example while it seemed to help in the previous one. What's the difference between the two, only the dataset?

I guess some dropout could be added before the final mapping, this could force the model to use the entire capacity of the final layer... I don't really know to be honest.

**Comment by csetraynor** [comment]
Yes, the only difference is the dataset. The one that helps is from a simulation using multivariate normal distributions. The one that doesn't help is from the census example dataset. I say more realistic because, at >70% dropout, I wasn't expecting it to help. Yes, some improvement from 10% to 50% dropout would be nice, still not improving, but using a 10% dropout is good enough for my goal. I will try to use it in the final mapping. Thank you for the suggestion.

**Comment by Optimox** [comment]
Closing this, feel free to reopen or propose changes in a PR if you feel like it. 
Thanks,

Seb
